File: atividade-10/atividade.go
package main

import (
	"fmt"
)

type AFN struct {
	states       map[string]struct{}
	startState   string
	acceptStates map[string]struct{}
}

func NewAFN() *AFN {
	return &AFN{
		states:       map[string]struct{}{"q0": {}, "q1": {}, "q2": {}},
		startState:   "q0",
		acceptStates: map[string]struct{}{"q2": {}},
	}
}

func (afn *AFN) isAccepted(input string) bool {
	currentStates := map[string]struct{}{afn.startState: {}}

	for _, char := range input {
		nextStates := make(map[string]struct{})
		for state := range currentStates {
			if state == "q0" {
				if char == '0' {
					nextStates["q1"] = struct{}{}
				}
			} else if state == "q1" {
				if char == '1' {
					nextStates["q2"] = struct{}{}
				} else if char == '0' {
					nextStates["q1"] = struct{}{}
				}
			} else if state == "q2" {
				nextStates["q2"] = struct{}{}
			}
		}
		currentStates = nextStates
	}

	for state := range currentStates {
		if _, exists := afn.acceptStates[state]; exists {
			return true
		}
	}
	return false
}

func main() {
	afn := NewAFN()
	fmt.Println(afn.isAccepted("010"))    
	fmt.Println(afn.isAccepted("0011"))   
	fmt.Println(afn.isAccepted("111"))    
	fmt.Println(afn.isAccepted("000"))    
	fmt.Println(afn.isAccepted("000111")) 
}

File: atividade-10/atividade.py
class AFN:
    def __init__(self):
        self.states = {'q0', 'q1', 'q2'}
        self.start_state = 'q0'
        self.accept_states = {'q2'}

    def is_accepted(self, string):
        current_states = {self.start_state}  

        for char in string:
            next_states = set()
            for state in current_states:
                if state == 'q0':
                    if char == '0':
                        next_states.add('q1')  
                elif state == 'q1':
                    if char == '1':
                        next_states.add('q2')  
                    elif char == '0':
                        next_states.add('q1')  
                elif state == 'q2':
                    next_states.add('q2')  
            
            current_states = next_states

        
        return any(state in self.accept_states for state in current_states)

if __name__ == '__main__':
    afn = AFN()
    print(afn.is_accepted("010"))   
    print(afn.is_accepted("0011"))  
    print(afn.is_accepted("111"))   
    print(afn.is_accepted("000"))   
    print(afn.is_accepted("000111")) 